src/include/platform/internal/GenericConfigurationManagerImpl.ipp

```diff
@@ -52,6 +53,162 @@ namespace chip {
 namespace DeviceLayer {
 namespace Internal {
 
+#if CHIP_USE_TRANSITIONAL_DEVICE_INSTANCE_INFO_PROVIDER
+template <class ConfigClass>
+class LegacyDeviceInstanceInfoProvider : public DeviceInstanceInfoProvider
+{
+
+public:
+    // GenericConfigurationManagerImpl will own a LegacyDeviceInstanceInfoProvider which
+    // *refers back to that GenericConfigurationManagerImpl*, due to how CRTP-based
+    // storage APIs are defined. This is a bit unclean, but only applicable to the
+    // transition path when `CHIP_USE_TRANSITIONAL_DEVICE_INSTANCE_INFO_PROVIDER` is true.
+    // This circular dependency is NOT needed by DeviceInstanceInfoProvider, but required
+    // to keep legacy code running.
+    LegacyDeviceInstanceInfoProvider(GenericConfigurationManagerImpl<ConfigClass> & configManager) :
+        mGenericConfigManager(configManager)
+    {}
+
+    CHIP_ERROR GetSerialNumber(MutableCharSpan & serialNumber) override;
+    CHIP_ERROR GetManufacturingDate(uint16_t & year, uint8_t & month, uint8_t & day) override;
+    CHIP_ERROR GetHardwareVersion(uint16_t & hardwareVersion) override;
+    CHIP_ERROR GetHardwareVersionString(MutableCharSpan & hardwareVersionString) override;
+    CHIP_ERROR GetRotatingDeviceIdUniqueId(MutableByteSpan & uniqueIdSpan) override;
+
+private:
+    GenericConfigurationManagerImpl<ConfigClass> & mGenericConfigManager;
+};
+
+template <class ConfigClass>
+CHIP_ERROR LegacyDeviceInstanceInfoProvider<ConfigClass>::GetSerialNumber(MutableCharSpan & serialNumber)
+{
+    ChipError err                                             = CHIP_NO_ERROR;
+    char valStr[mGenericConfigManager.kMaxSerialNumberLength] = { 0 };
+    size_t valStrLen                                          = 0;
+
+    err = mGenericConfigManager.ReadConfigValueStr(ConfigClass::kConfigKey_SerialNum, valStr, sizeof(valStr), valStrLen);
+
+#ifdef CHIP_DEVICE_CONFIG_TEST_SERIAL_NUMBER
+    if (CHIP_DEVICE_CONFIG_TEST_SERIAL_NUMBER[0] != 0 && err == CHIP_DEVICE_ERROR_CONFIG_NOT_FOUND)
+    {
+        ReturnErrorCodeIf(sizeof(CHIP_DEVICE_CONFIG_TEST_SERIAL_NUMBER) > serialNumber.size(), CHIP_ERROR_BUFFER_TOO_SMALL);
+        memcpy(valStr, CHIP_DEVICE_CONFIG_TEST_SERIAL_NUMBER, sizeof(CHIP_DEVICE_CONFIG_TEST_SERIAL_NUMBER));
+        valStrLen = sizeof(CHIP_DEVICE_CONFIG_TEST_SERIAL_NUMBER) - 1;
+    }
+#endif // CHIP_DEVICE_CONFIG_TEST_SERIAL_NUMBER
+
+    ReturnErrorOnFailure(err);
+    // compare to argument's size - 1 due to null terminating at last byte
+    ReturnErrorCodeIf(valStrLen > serialNumber.size() - 1, CHIP_ERROR_BUFFER_TOO_SMALL);
```

Comment:
Where is the null-terminator added? I think the point of using Spans is not using the null-terminator. Otherwise, we have two alternative methods of telling what the string length is which may easily get out of sync.